File: react/src/classes/ClientCommands/MotorConfigurationCommand.ts
import { motorIdValidator } from "@/validators/motorIdValidator";
import { z } from "zod";
import {
  AbstractClientCommand,
  TClientCommandExecuteParam,
} from "./AbstractClientCommand";
import { setAvailableColors, setSelectedColor } from "@/store/colorSlice";
import { setActiveMotorId, setMotors } from "@/store/motorSlice";
import { Motor } from "@/classes/socketClients/Motor";

export const motorConfigurationCommandPayloadItem = z.object({
  motorId: motorIdValidator,
  name: z.string(),
  colorCode: z.string().regex(/#[0-9a-fA-F]{6}/),
  isRunning: z.boolean(),
  direction: z.literal("up").or(z.literal("down")),
});

export const motorConfigurationCommand = z.object({
  command: z.literal("motorConfiguration"),
  payload: z.array(motorConfigurationCommandPayloadItem),
});

export type TMotorConfigurationItem = z.infer<
  typeof motorConfigurationCommandPayloadItem
>;

export type TMotorConfigurationCommand = z.infer<
  typeof motorConfigurationCommand
>;

export class MotorConfigurationCommand extends AbstractClientCommand<TMotorConfigurationCommand> {
  execute({
    payload,
    dispatch,
    ws,
  }: TClientCommandExecuteParam<TMotorConfigurationCommand>) {
    if (payload.length < 1) return;
    dispatch(setAvailableColors(payload));
    dispatch(setSelectedColor(payload[0]));
    dispatch(setMotors(payload.map((config) => new Motor(ws, config))));
    dispatch(setActiveMotorId(payload[0].motorId));
    console.log("MotorConfigurationCommand::execute()", payload);
  }
}

File: src/inc/commands/client/MotorConfigurationCommand.cpp
#include "./MotorConfigurationCommand.h"

MotorConfigurationCommand::MotorConfigurationCommand() {}

JsonDocument MotorConfigurationCommand::toJson()
{
    JsonDocument doc;

    doc["command"] = "motorConfiguration";
    JsonArray payload = doc["payload"].to<JsonArray>();

    for (const auto &motor : motorConfig)
    {
        if (!motor.active)
            continue;
        JsonObject obj = payload.add<JsonObject>();

        int pin = stepper[motor.id]->getDirectionPin();
        delay(100);

        obj["motorId"] = motor.id;
        obj["name"] = motor.name;
        obj["isRunning"] = stepper[motor.id]->isRunning();
        obj["direction"] = digitalRead(pin) ? "up" : "down";
        obj["colorCode"] = motor.colorCode;
    }

    return doc;
}